tasm-lib/src/arithmetic/u128/overflowing_add.rs

```diff
@@ -0,0 +1,289 @@
+use triton_vm::prelude::*;
+
+use crate::data_type::DataType;
+use crate::library::Library;
+use crate::traits::basic_snippet::BasicSnippet;
+
+#[derive(Clone, Debug, Copy)]
+pub struct OverflowingAddU128;
+
+impl BasicSnippet for OverflowingAddU128 {
+    fn entrypoint(&self) -> String {
+        "tasmlib_arithmetic_u128_overflowing_add".to_string()
+    }
+
+    fn inputs(&self) -> Vec<(DataType, String)> {
+        vec![
+            (DataType::U128, "lhs".to_owned()),
+            (DataType::U128, "rhs".to_owned()),
+        ]
+    }
+
+    fn outputs(&self) -> Vec<(DataType, String)> {
+        vec![
+            (DataType::U128, "sum".to_owned()),
+            (DataType::Bool, "overflow".to_owned()),
+        ]
+    }
+
+    /// Four top elements of stack are assumed to be valid u32s. So to have
+    /// a value that's less than 2^32.
+    fn code(&self, _library: &mut Library) -> Vec<LabelledInstruction> {
+        let entrypoint = self.entrypoint();
+
+        triton_asm!(
+            // BEFORE: _ rhs_3 rhs_2 rhs_1 rhs_0 lhs_3 lhs_2 lhs_1 lhs_0
+            // AFTER:  _ sum_3 sum_2 sum_1 sum_0
+            {entrypoint}:
+                swap 1 swap 4
+                add
+                // _ rhs_3 rhs_2 rhs_1 lhs_1 lhs_3 lhs_2 (lhs_0 + rhs_0)
+
+                split
+                // _ rhs_3 rhs_2 rhs_1 lhs_1 lhs_3 lhs_2 (lhs_0 + rhs_0)_hi (lhs_0 + rhs_0)_lo
+
+                // rename:
+                // _ rhs_3 rhs_2 rhs_1 lhs_1 lhs_3 lhs_2 carry_1 sum_0
+
+                swap 4
+                // _ rhs_3 rhs_2 rhs_1 sum_0 lhs_3 lhs_2 carry_1 lhs_1
+
+                add
+                // _ rhs_3 rhs_2 rhs_1 sum_0 lhs_3 lhs_2 lhs_1'
+
+                swap 1 swap 4
+                // _ rhs_3 rhs_2 lhs_2 sum_0 lhs_3 lhs_1' rhs_1
+
+                add
+                // _ rhs_3 rhs_2 lhs_2 sum_0 lhs_3 (lhs_1' + rhs_1)
+
+                split
+                // _ rhs_3 rhs_2 lhs_2 sum_0 lhs_3 carry_2 sum_1
+
+                swap 4
+                // _ rhs_3 rhs_2 sum_1 sum_0 lhs_3 carry_2 lhs_2
+
+                add
+                // _ rhs_3 rhs_2 sum_1 sum_0 lhs_3 lhs_2'
+
+                swap 1 swap 4
+                // _ rhs_3 lhs_3 sum_1 sum_0 lhs_2' rhs_2
+
+                add
+                // _ rhs_3 lhs_3 sum_1 sum_0 (lhs_2' + rhs_2)
+
+                split
+                // _ rhs_3 lhs_3 sum_1 sum_0 carry_3 sum_2
+
+                swap 4
+                // _ rhs_3 sum_2 sum_1 sum_0 carry_3 lhs_3
+
+                add
+                // _ rhs_3 sum_2 sum_1 sum_0 lhs_3'
+
+                dup 4
+                // _ rhs_3 sum_2 sum_1 sum_0 lhs_3' rhs_3
+
+                add
+                // _ rhs_3 sum_2 sum_1 sum_0 (lhs_3' + rhs_3)
+
+                split
+                // _ rhs_3 sum_2 sum_1 sum_0 overflow sum_3
+
+                swap 5
+                pop 1
+                // _ sum_3 sum_2 sum_1 sum_0 overflow
+
+                return
+        )
+    }
```

Comment:
As is usually the case, using `pick` and `place` is more efficient than `swap`. The snippet below uses 18 instructions instead of the suggested 23.

```rust
fn code(&self, _: &mut Library) -> Vec<LabelledInstruction> {
    triton_asm!(
        // BEFORE: _ rhs_3 rhs_2 rhs_1 rhs_0 lhs_3 lhs_2 lhs_1 lhs_0
        // AFTER:  _ sum_3 sum_2 sum_1 sum_0 is_overflow
        //                                   ^^^^^^^^^^^
        //             don't forget to adapt signature when copy-pasting
        {self.entrypoint()}:
            pick 4
            // _ rhs_3 rhs_2 rhs_1 lhs_3 lhs_2 lhs_1 lhs_0 rhs_0

            add
            split
            // _ rhs_3 rhs_2 rhs_1 lhs_3 lhs_2 lhs_1 (lhs_0 + rhs_0)_hi (lhs_0 + rhs_0)_lo
            // _ rhs_3 rhs_2 rhs_1 lhs_3 lhs_2 lhs_1 carry_1            sum_0

            place 7
            pick 4
            // _ sum_0 rhs_3 rhs_2 lhs_3 lhs_2 lhs_1 carry_1 rhs_1

            add
            add
            split
            // _ sum_0 rhs_3 rhs_2 lhs_3 lhs_2 carry_2 sum_1

            place 6
            pick 3
            // _ sum_1 sum_0 rhs_3 lhs_3 lhs_2 carry_2 rhs_2

            add
            add
            split
            // _ sum_1 sum_0 rhs_3 lhs_3 carry_3 sum_2

            place 5
            // _ sum_2 sum_1 sum_0 rhs_3 lhs_3 carry_3

            add
            add
            split
            // _ sum_2 sum_1 sum_0 carry_4 sum_3

            place 4
            // _ sum_3 sum_2 sum_1 sum_0 carry_4
            // _ sum_3 sum_2 sum_1 sum_0 is_overflow

            return
    )
}
```